### PR TITLE
session,expression: Support IS_USED_LOCK function

### LINKED ISSUE #44493: Support `IS_USED_LOCK()` and `IS_FREE_LOCK()` functions
## Enhancement

See also 
- #14994
- https://dev.mysql.com/doc/refman/8.0/en/locking-functions.html#function_is-used-lock
- https://dev.mysql.com/doc/refman/8.0/en/locking-functions.html#function_is-free-lock

This would improve compatibility with MySQL. Looks like Magento might benefit from this.

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #44493 

Problem Summary:

### What is changed and how it works?

TiDB, Session 1
```
sql> select is_used_lock('foo');
+---------------------+
| is_used_lock('foo') |
+---------------------+
|                NULL |
+---------------------+
1 row in set (0.0037 sec)

sql> select get_lock('foo', 10);
+---------------------+
| get_lock('foo', 10) |
+---------------------+
|                   1 |
+---------------------+
1 row in set (0.0032 sec)

sql> select is_used_lock('foo');
+---------------------+
| is_used_lock('foo') |
+---------------------+
|       2199023255557 |
+---------------------+
1 row in set (0.0005 sec)

sql> select connection_id();
+-----------------+
| connection_id() |
+-----------------+
|   2199023255557 |
+-----------------+
1 row in set (0.0004 sec)
```

TiDB, Session 2
```
sql> select is_used_lock('foo');
+---------------------+
| is_used_lock('foo') |
+---------------------+
|                   1 |
+---------------------+
1 row in set (1.0120 sec)

sql> select is_used_lock('bar');
+---------------------+
| is_used_lock('bar') |
+---------------------+
|                NULL |
+---------------------+
1 row in set (0.0033 sec)
```

MySQL
```
sql> select is_used_lock('foo');
+---------------------+
| is_used_lock('foo') |
+---------------------+
|                NULL |
+---------------------+
1 row in set (0.0007 sec)

sql> select get_lock('foo',10);
+--------------------+
| get_lock('foo',10) |
+--------------------+
|                  1 |
+--------------------+
1 row in set (0.0007 sec)

sql> select is_used_lock('foo');
+---------------------+
| is_used_lock('foo') |
+---------------------+
|                  36 |
+---------------------+
1 row in set (0.0007 sec)

sql> select connection_id();
+-----------------+
| connection_id() |
+-----------------+
|              36 |
+-----------------+
1 row in set (0.0006 sec)

sql> select version();
+-----------+
| version() |
+-----------+
| 8.0.33    |
+-----------+
1 row in set (0.0007 sec)
```
### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support for the `IS_USED_LOCK()` and `IS_FREE_LOCK()` functions was added
```

### CODE REVIEW

**Comment by dveeden** [comment]
The problem with the current implementation is that `lock.owner` is only available to the current session.  For other sessions (possibly on other nodes) this basically tries to get and release a transaction to see if a lock exists. The problem with this is that it doesn't return information on who owns the lock. Now it always returns `1` instead of the proper process ID because of that.

* Could we check lock metadata instead of lock/release to check if a lock exists?
* We could add another column to the `mysql.advisory_locks` table to record the owner. However as the transaction never gets commited it would be difficult to get the record from other sessions/nodes.

**Comment by dveeden** [comment]
The `evalInt()` now does a `int64()` on the owner pid before returning it. So sessions with an ID above `MaxInt64` would likely have issues. Not sure how to properly return higher values.

File: session/advisory_locks.go [hunk 1]
```diff
@@ -88,3 +89,23 @@ func (a *advisoryLock) GetLock(lockName string, timeout int64) error {
...
+	a.ctx = kv.WithInternalSourceType(a.ctx, kv.InternalTxnOthers)
+	_, err := a.session.ExecuteInternal(a.ctx, "SET innodb_lock_wait_timeout = 1")
+	if err != nil {
+		return err
+	}
+	_, err = a.session.ExecuteInternal(a.ctx, "BEGIN PESSIMISTIC")
+	if err != nil {
+		return err
+	}
+	_, err = a.session.ExecuteInternal(a.ctx, "INSERT INTO mysql.advisory_locks (lock_name) VALUES (%?)", lockName)
```
[the remark this answers is not in the document]
Comment (dveeden) [reply session/advisory_locks.go:105]: Done

**Comment by dveeden** [comment]
> Maybe also add [IS_FREE_LOCK()](https://dev.mysql.com/doc/refman/5.7/en/locking-functions.html#function_is-free-lock) as an inverse of `IS_USED_LOCK()`?

Done in edb9729aefc44d6beb16837adc6d4b1e4982f84b

MySQL
```
sql> select is_free_lock('test');
+----------------------+
| is_free_lock('test') |
+----------------------+
|                    1 |
+----------------------+
1 row in set (0.0004 sec)

sql> select get_lock('test',1);
+--------------------+
| get_lock('test',1) |
+--------------------+
|                  1 |
+--------------------+
1 row in set (0.0007 sec)

sql> select is_free_lock('test');
+----------------------+
| is_free_lock('test') |
+----------------------+
|                    0 |
+----------------------+
1 row in set (0.0064 sec)

sql> select version();
+-----------+
| version() |
+-----------+
| 8.0.33    |
+-----------+
1 row in set (0.0007 sec)
```

TiDB
```sql>  select is_free_lock('test');
+----------------------+
| is_free_lock('test') |
+----------------------+
|                    1 |
+----------------------+
1 row in set (0.0042 sec)

sql> select get_lock('test',1);
+--------------------+
| get_lock('test',1) |
+--------------------+
|                  1 |
+--------------------+
1 row in set (0.0026 sec)

sql> select is_free_lock('test');
+----------------------+
| is_free_lock('test') |
+----------------------+
|                    0 |
+----------------------+
1 row in set (0.0010 sec)

```

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**